File: tests/test_group_system.py
import pytest
from balanced_group_system import GroupSystem

@pytest.fixture
def group_system():
  return GroupSystem(['Alice', 'Bob', 'Charlie', 'David', 'Eve'])

def test_init(group_system):
  assert group_system.members == ['Alice', 'Bob', 'Charlie', 'David', 'Eve']
  assert group_system.group_history == []

def test_add_member(group_system):
  group_system.add_member('Frank')
  assert group_system.members == ['Alice', 'Bob', 'Charlie', 'David', 'Eve', 'Frank']

def test_remove_member(group_system):
  group_system.remove_member('Bob')
  assert group_system.members == ['Alice', 'Charlie', 'David', 'Eve']

def test_create_groups(group_system):
  group_list1 = [['Alice', 'Bob'], ['Charlie', 'David', 'Eve']]
  group_list2 = [['Alice', 'Bob'], ['Charlie'], ['David', 'Eve']]
  group_system.create_groups(group_list1)
  group_system.create_groups(group_list2)
  assert group_system.group_history == [group_list1, group_list2]

def test_create_validate_groups(group_system):
  # Test case: Member not in groupSystem.members
  with pytest.raises(KeyError) as exc_info:
    group_system.create_and_validate_groups([['Alice', 'Bob'], ['Charlie', 'David', 'Frank']])
    assert str(exc_info.value) == "Element 'Frank' was not in members list"
  
  # Test case: Duplicate member within a group
  with pytest.raises(ValueError) as exc_info:
    group_system.create_and_validate_groups([['Alice', 'Bob'], ['Charlie', 'Charlie']])
    assert str(exc_info.value) == "Duplicate members found in group_list"

File: balanced_group_system/test_group_system.cpp
#include "group_system.hpp"
#include <iostream>
#include <vector>
#include <cassert>
#include <stdexcept>

void test_init() {
    std::vector<std::string> members = {"Alice", "Bob", "Charlie", "David", "Eve"};
    GroupSystem gs(members);
    assert(gs.members == members);
    assert(gs.group_history == std::vector<std::vector<std::vector<std::string>>>(0));
}

void test_add_member() {
    std::vector<std::string> members = {"Alice", "Bob", "Charlie", "David", "Eve"};
    GroupSystem gs(members);
    gs.add_member("Frank");
    assert(gs.members.size() == 6);
}

void test_remove_member() {
    std::vector<std::string> members = {"Alice", "Bob", "Charlie", "David", "Eve"};
    GroupSystem gs(members);
    gs.remove_member("Bob");
    assert(gs.members.size() == 4);
}

void test_create_groups() {
    std::vector<std::string> members = {"Alice", "Bob", "Charlie", "David", "Eve"};
    GroupSystem gs(members);
    std::vector<std::vector<std::string>> group_list1 = {{"Alice", "Bob"}, {"Charlie", "David", "Eve"}};
    std::vector<std::vector<std::string>> group_list2 = {{"Alice", "Bob"}, {"Charlie"}, {"David", "Eve"}};
    gs.create_groups(group_list1);
    gs.create_groups(group_list2);
    assert(gs.group_history == std::vector<std::vector<std::vector<std::string>>>({group_list1, group_list2}));
}

void test_create_validate_groups() {
    std::vector<std::string> members = {"Alice", "Bob", "Charlie", "David", "Eve"};
    GroupSystem gs(members);
    // Test case: Member not in groupSystem.members
    try{
        gs.create_and_validate_groups({{"Alice", "Bob"}, {"Charlie", "David", "Frank"}});
        assert(false);
    } catch(const std::runtime_error& error){
        assert(std::string(error.what()) == "Member not found");
    }
    
    // Test case: Duplicate member within a group
    try {
        gs.create_and_validate_groups({{"Alice", "Bob"}, {"Charlie", "Charlie"}});
    } catch(const std::runtime_error& error){
        assert(std::string(error.what()) == "Duplicate members found");
    }
}

int main() {
    test_init();
    test_add_member();
    test_remove_member();
    test_create_groups();
    test_create_validate_groups();
}
